helix-term/src/commands.rs

```diff
@@ -500,80 +481,79 @@ fn goto_window_bottom(cx: &mut Context) {
 fn move_next_word_start(cx: &mut Context) {
     let count = cx.count();
     let (view, doc) = current!(cx.editor);
+    let text = doc.text().slice(..);
 
-    doc.set_selection(
-        view.id,
-        doc.selection(view.id)
-            .clone()
-            .min_width_1(doc.text().slice(..))
-            .transform(|range| movement::move_next_word_start(doc.text().slice(..), range, count)),
-    );
+    let selection = doc
+        .selection(view.id)
+        .clone()
+        .min_width_1(text)
+        .transform(|range| movement::move_next_word_start(text, range, count));
+    doc.set_selection(view.id, selection);
 }
 
 fn move_prev_word_start(cx: &mut Context) {
     let count = cx.count();
     let (view, doc) = current!(cx.editor);
-    doc.set_selection(
-        view.id,
-        doc.selection(view.id)
-            .clone()
-            .min_width_1(doc.text().slice(..))
-            .transform(|range| movement::move_prev_word_start(doc.text().slice(..), range, count)),
-    );
+    let text = doc.text().slice(..);
+
+    let selection = doc
+        .selection(view.id)
+        .clone()
+        .min_width_1(text)
+        .transform(|range| movement::move_prev_word_start(text, range, count));
+    doc.set_selection(view.id, selection);
 }
 
 fn move_next_word_end(cx: &mut Context) {
     let count = cx.count();
     let (view, doc) = current!(cx.editor);
-    doc.set_selection(
-        view.id,
-        doc.selection(view.id)
-            .clone()
-            .min_width_1(doc.text().slice(..))
-            .transform(|range| movement::move_next_word_end(doc.text().slice(..), range, count)),
-    );
+    let text = doc.text().slice(..);
+
+    let selection = doc
+        .selection(view.id)
+        .clone()
```

Comment:
Why do we need clone for quite a few of these?

Reply:
Because with the current API you can't directly mutate the document's selection.  You have to make a new selection and then set it.

Before this PR the `transform()` method cloned internally and returned the clone, so the old code was cloning as well, just not explicitly.  Now `transform()` and friends are consuming, so they don't clone and you can chain transformations together without additional clones.

(Also, in an earlier version of this PR I added direct mutation access to the document's selection, and used that to avoid allocations in these places.  But @archseer preferred the more functional API, so I changed it back.)

In practice, I doubt the clone is a problem for command performance.  It's a single allocation + memcopy (or should be, at least).  And almost every command is going to loop over all ranges anyway, so the memcopy is cheap compared to the rest of the work being done.